### PR TITLE
IGNITE-15298 Partitions raft snapshots

### PR DESCRIPTION
**Pull request by SammyVimes**
https://issues.apache.org/jira/browse/IGNITE-15298

### CODE REVIEW

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+
+        for (ClusterService service : cluster)
+            service.stop();
+    }
+
+    /**
+     * Test parameters for {@link #testSnapshot}.
+     */
+    private static class TestData {
+        /** Delete raft group folder. */
```
Comment (sashapolo) [inline modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:116]: please add more verbose description, it is not clear, how do these flags influence the tests

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    /**
+     * Interacts with raft group before the follower is stopped.
+     *
+     * @param service Raft group service.
+     * @throws Exception If failed.
+     */
+    public abstract void doBeforeStop(RaftGroupService service) throws Exception;
+
+    /**
+     * Interacts with raft group after the follower is stopped.
```
Comment (sashapolo) [inline modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:222]:
```suggestion
     * Interacts with the raft group after a follower is stopped.
```

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+     *
+     * @param service Raft group service.
+     * @throws Exception If failed.
+     */
+    public abstract void doAfterStop(RaftGroupService service) throws Exception;
+
+    /**
+     * Interacts with raft group after the leader has captured a snapshot.
```
Comment (sashapolo) [inline modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:230]:
```suggestion
     * Interacts with a raft group after the leader has captured a snapshot.
```

Comment (sashapolo) [reply modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:230]: what do you mean by "interacts"?

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+
+        BooleanSupplier closure = snapshotCheckClosure(restarted, testData.interactAfterSnapshot);
+
+        boolean success = waitForCondition(closure, 10_000);
+
+        assertTrue(success);
+    }
+
+    /**
+     * Interacts with raft group before the follower is stopped.
```
Comment (sashapolo) [inline modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:214]:
```suggestion
     * Interacts with the raft group before a follower is stopped.
```

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+     *
+     * @param service Raft group service.
+     * @throws Exception If failed.
+     */
+    public abstract void doBeforeStop(RaftGroupService service) throws Exception;
```
Comment (sashapolo) [inline modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:219]: I think that `beforeFollowerStop`, `afterFollowerStop`, `afterSnapshot` might be better names for these methods

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+
+    /**
+     * Creates a closure that will be executed periodically to check if the snapshot and (conditionally on the
+     * {@link TestData#interactAfterSnapshot}) the raft log was successfully restored by the follower node.
+     *
+     * @param restarted Restarted follower node.
+     * @param interactedAfterSnapshot {@code true} whether raft group was interacted with after the snapshot operation.
+     * @return Closure.
+     */
+    public abstract BooleanSupplier snapshotCheckClosure(JRaftServerImpl restarted, boolean interactedAfterSnapshot);
```
Comment (sashapolo) [inline modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:245]: Actually I think that all these abstract methods are a sign of over-engineering. Do we really need an abstract class for these two tests? I think it is easier to copy-paste the test itself....

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ITTablePersistenceTest.java [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.ignite.raft.client.service.ITAbstractListenerSnapshotTest;
+import org.apache.ignite.raft.client.service.RaftGroupListener;
+import org.apache.ignite.raft.client.service.RaftGroupService;
+
+/**
+ * Persistent (rocksdb-based) partitions raft group snapshots tests.
+ */
+public class ITTablePersistenceTest extends ITAbstractListenerSnapshotTest<PartitionListener> {
+    /** */
+    public static SchemaDescriptor SCHEMA = new SchemaDescriptor(UUID.randomUUID(),
```
Comment (sashapolo) [inline modules/table/src/integrationTest/java/org/apache/ignite/distributed/ITTablePersistenceTest.java:47]:
```suggestion
    private static final SchemaDescriptor SCHEMA = new SchemaDescriptor(UUID.randomUUID(),
```

File: modules/table/src/integrationTest/java/org/apache/ignite/distributed/ITTablePersistenceTest.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        1,
+        new Column[] {new Column("key", NativeTypes.INT64, false)},
+        new Column[] {new Column("value", NativeTypes.INT64, false)}
+    );
+
+    /** */
+    private static final Row firstKey = createKeyRow(0);
```
Comment (sashapolo) [inline modules/table/src/integrationTest/java/org/apache/ignite/distributed/ITTablePersistenceTest.java:54]: shouldn't these variables be named in upper case?

File: modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java (hunk 6 above)
Comment (SammyVimes) [reply modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:245]: Tests are basically the same, any change in the raft group API would require changes in two places

Comment (sashapolo) [reply modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:251]: Yes, but all these abstract methods show that they are not really the same, only the overall test scenario

Comment (SammyVimes) [reply modules/raft/src/integrationTest/java/org/apache/ignite/raft/client/service/ITAbstractListenerSnapshotTest.java:251]: I think raft set up, stop and restart operations are too complicated to copy-paste them. I don't see how abstract tests can hurt us

File: modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/RocksUtils.java [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+        catch (RocksDBException e) {
+            throw new IgniteInternalException(e);
+        }
+    }
+
+    /**
+     * BiConsumer that can throw {@link RocksDBException}.
+     */
+    @FunctionalInterface
+    public interface RocksBiConsumer {
```
Comment (sashapolo) [inline modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/RocksUtils.java:131]: let's extract these classes to the upper level

File: modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/RocksUtils.java [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    /**
+     * Creates an SST file for the column family.
+     *
+     * @param columnFamily Column family.
+     * @param snapshot Point-in-time snapshot.
+     * @param path Directory to put the SST file in.
+     * @param snapshotExecutor Snapshot executor.
+     * @return Future representing pending completion of the operation.
+     */
+    public static CompletableFuture<Void> createSstFile(
```
Comment (sashapolo) [inline modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/RocksUtils.java:45]: I would suggest making this method synchronous and dispatch it onto executors on the caller's side

**Pull request closed without merging**